### PR TITLE
WIP: Patch queue: refine optimizer estimations, refine rules, try different cost and metadata propagation

### PR DESCRIPTION
**Pull request by vlsi**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java [hunk 1]
```diff
@@ -101,44 +101,21 @@ public static EnumerableHashJoin create(
...
-    default:
-      if (RelNodes.COMPARATOR.compare(left, right) > 0) {
-        rowCount = RelMdUtil.addEpsilon(rowCount);
-      }
-    }
-
-    // Cheaper if the smaller number of rows is coming from the LHS.
-    // Model this by adding L log L to the cost.
+    RelOptCostFactory costFactory = planner.getCostFactory();
     final double rightRowCount = right.estimateRowCount(mq);
```
Comment (rubenada) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java]: First of all, thanks for you huge effort on this PR, @vlsi 
I have a small comment here, according to RelNode#estimateRowCount javadoc: _"don't call this method directly. Instead, use RelMetadataQuery#getRowCount, which gives plugins a chance to override the rel's default ideas about row count"._ I wonder if we could use this PR to replace:
```
final double rightRowCount = right.estimateRowCount(mq);
final double leftRowCount = left.estimateRowCount(mq);
```
with
```
final double rightRowCount = mq.getRowCount(right);
final double leftRowCount = mq.getRowCount(left);
```
which seems more correct, according to that javadoc comment. We could do that in EnumerableHashJoin, EnumerableMergeJoin, EnumerableNestedLoopJoin, EnumerableBatchNestedLoopJoin and Correlate (which were already impacted by the PR). What do you think?

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java]: Thanks. It makes sense indeed. Will replace shortly

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableTableScan.java [hunk 2]
```diff
@@ -84,15 +90,59 @@ public static EnumerableTableScan create(RelOptCluster cluster,
...
+      return true;
+    }
+    // Struct fields are not supported in EnumerableTableScan
+    for (RelDataTypeField field : relOptTable.getRowType().getFieldList()) {
+      boolean unsupportedType = false;
+      switch (field.getType().getSqlTypeName()) {
+      case ARRAY:
+        unsupportedType = supportArray;
+        break;
+      case MAP:
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableTableScan.java]: Better move the type decision to `SqlTypeUtil`.

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableTableScan.java]: I guess this is already in master.
Will rebase this PR.

**Pull request closed without merging**